lib/view_component/compiler.rb

```diff
@@ -0,0 +1,214 @@
+# frozen_string_literal: true
+
+module ViewComponent
+  class Compiler
+    def initialize(component_class)
+      @component_class = component_class
+    end
+
+    def compiled?
+      CompileCache.compiled?(component_class)
+    end
+
+    def compile(raise_errors: false)
+      return if compiled?
+
+      if template_errors.present?
+        raise ViewComponent::TemplateError.new(template_errors) if raise_errors
+        return false
+      end
+
+      if component_class.instance_methods(false).include?(:before_render_check)
+        ActiveSupport::Deprecation.warn(
+          "`before_render_check` will be removed in v3.0.0. Use `before_render` instead."
+        )
+      end
+
+      # Remove any existing singleton methods,
+      # as Ruby warns when redefining a method.
+      component_class.remove_possible_singleton_method(:collection_parameter)
+      component_class.remove_possible_singleton_method(:collection_counter_parameter)
+      component_class.remove_possible_singleton_method(:counter_argument_present?)
+
+      component_class.define_singleton_method(:collection_parameter) do
+        if provided_collection_parameter
+          provided_collection_parameter
+        else
+          name.demodulize.underscore.chomp("_component").to_sym
+        end
+      end
+
+      component_class.define_singleton_method(:collection_counter_parameter) do
+        "#{collection_parameter}_counter".to_sym
+      end
+
+      component_class.define_singleton_method(:counter_argument_present?) do
+        instance_method(:initialize).parameters.map(&:second).include?(collection_counter_parameter)
+      end
+
+      component_class.validate_collection_parameter! if raise_errors
+
+      templates.each do |template|
+        # Remove existing compiled template methods,
+        # as Ruby warns when redefining a method.
+        method_name = call_method_name(template[:variant])
+        component_class.send(:undef_method, method_name.to_sym) if component_class.instance_methods.include?(method_name.to_sym)
+
+        component_class.class_eval <<-RUBY, template[:path], -1
+          def #{method_name}
+            @output_buffer = ActionView::OutputBuffer.new
+            #{compiled_template(template[:path])}
+          end
+        RUBY
+      end
+
+      define_render_template_for
+
+      CompileCache.register(component_class)
+    end
+
+    private
+
+    attr_reader :component_class
+
+    def define_render_template_for
```

Comment:
I think this might be the only introduction of new code here, but this gives us a small performance boost by letting us avoid having to call `send` to render the correct template in the component.